pilot/pkg/networking/core/cluster_waypoint.go

```diff
@@ -312,18 +327,62 @@ func (cb *ClusterBuilder) buildWaypointInboundVIP(proxy *model.Proxy, svcs map[h
 	return clusters
 }
 
+// buildWaypointInnerConnectOriginate creates a cluster that innitiate inner HBONE tunnel of double-HBONE.
+// InnerConnectOrigiante cluster is responsible for creating the inner CONNECT tunnel and forwarding to an internal listener that
+// will wrap the traffic into outer HBONE tunnel.
+func (cb *ClusterBuilder) buildWaypointInnerConnectOriginate(proxy *model.Proxy, push *model.PushContext) *cluster.Cluster {
+	// Normally for clusters that just redirect to internal listeners we would use util.DefaultInternalUpstreamTransportSocket.
+	// util.DefaultInternalUpstreamTransportSocket is a InternalUpstreamTransport socket wrapping a RawBufferTransport socket.
+	// For double HBONE we want something slightly different - we still need to use InternalUpstreamTransport socket because
+	// we redirect to an internal listener, but instead of keeping data as-is we want to encrypt it using TLS, so instead of the
+	// RawBufferTransport we want to wrap a TLS transport socket.
+	tlsCtx := buildCommonConnectTLSContext(proxy, push)
+	sec_model.EnforceCompliance(tlsCtx)
+	transportSocket := util.InternalUpstreamTransportSocket("internal_upstream_with_tls", &core.TransportSocket{
+		Name: "tls",
+		ConfigType: &core.TransportSocket_TypedConfig{TypedConfig: protoconv.MessageToAny(&tlsv3.UpstreamTlsContext{
+			CommonTlsContext: tlsCtx,
+		})},
+	})
+
+	c := &cluster.Cluster{
+		Name:                 InnerConnectOriginate,
+		ClusterDiscoveryType: &cluster.Cluster_Type{Type: cluster.Cluster_STATIC},
+		CircuitBreakers: &cluster.CircuitBreakers{
+			Thresholds: []*cluster.CircuitBreakers_Thresholds{getDefaultCircuitBreakerThresholds()},
+		},
+		LoadAssignment: &endpoint.ClusterLoadAssignment{
+			ClusterName: InnerConnectOriginate,
+			Endpoints:   util.BuildInternalEndpoint(OuterConnectOriginate, nil),
+		},
+		TypedExtensionProtocolOptions: h2connectUpgrade(),
+		TransportSocket:               transportSocket,
+	}
+
+	c.AltStatName = util.DelimitedStatsPrefix(InnerConnectOriginate)
+
+	return c
+}
+
+// buildWaypointOuterConnectOriginate creates a cluster that finishes wrapping traffic in double HBONE by finalizing outer HBONE tunnel.
+// It's basically equivalent to the regular waypoint ConnectOriginate cluster and does the same thing, the only real difference is that
+// it wraps the data already wrapped into a CONNECT once.
+func (cb *ClusterBuilder) buildWaypointOuterConnectOriginate(proxy *model.Proxy, push *model.PushContext) *cluster.Cluster {
```

Comment:
Do we need to do anything differently w.r.t the connection pool of this outer HBONE tunnel?

Reply:
Does that also apply to the inner tunnel too? My concern is that maybe we pin ourselves to one particular backend in the remote cluster per waypoint (until that connection dies). We can probably optimize later but should document it

Reply:
> I also tried to set max_requests_per_connection to 1, to effecintely disable connection pooling on both inner and outer cluster. When I did that, I saw different backends being used in the remote network, but it definitely didn't create a new connection for each request - I don't understand what is the algorithm there.

Is it worth doing this for now and then do some tuning in the next release?

Reply:
I think I figured out the current behavior when we disable pooling in the inner/outer connect originate cluster. Basically we establish a new connection after some time of inactivity. That also means that disabling pooling in inner/outer clusters is basically useless when it actually matters (i.e., we laser the same backend when we are under load and don't do that when we are not - that's the oposite of what we want really).

Reply:
Filed https://github.com/istio/istio/issues/58039 to followup on a proper solution to avoid lasering.